### PR TITLE
Bind aux socket to get a port

### PR DESCRIPTION
**Pull request by maolson-msft**
For each CXPLAT_SOCKET, bind an auxiliary Winsock socket to acquire and reserve a port from tcpip.sys.

### CODE REVIEW

File: src/platform/datapath_raw_socket.c [hunk 1]
```diff
@@ -79,20 +76,91 @@ CxPlatTryAddSocket(
...
     CXPLAT_HASHTABLE_LOOKUP_CONTEXT Context;
     CXPLAT_HASHTABLE_ENTRY* Entry;
 
+    //
+    // Get (and reserve) a transport layer port from tcpip.sys by binding an auxiliary socket.
+    //
+
+    Socket->AuxSocket =
+        WSASocketW(
+            AF_INET6,
```
Comment (csujedihy) [inline src/platform/datapath_raw_socket.c:90]: A comment explaining why a IPv6 socket is used would help.

Comment (nibanks) [reply src/platform/datapath_raw_socket.c:90]: Can we use the normal `socket` function to keep this code xplat?

File: src/platform/datapath_raw_socket.c [hunk 2]
```diff
@@ -79,20 +76,91 @@ CxPlatTryAddSocket(
...
     CxPlatRwLockAcquireExclusive(&Pool->Lock);
 
-    if (Socket->LocalAddress.Ipv4.sin_port == 0) {
-        Socket->LocalAddress.Ipv4.sin_port = CxPlatByteSwapUint16(Pool->NextLocalPort);
-        Pool->NextLocalPort++;
-        if (Pool->NextLocalPort == 0) {
-            Pool->NextLocalPort = HARDCODED_SOCK_POOL_START_PORT;
-        }
+    Result =
+        bind(
```
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:129]: When `Socket-Connected` is `TRUE` we need to call `connect` as well.

File: src/platform/datapath_raw_socket.c [hunk 3]
```diff
@@ -105,7 +173,11 @@ CxPlatTryAddSocket(
     if (Success) {
         CxPlatHashtableInsert(&Pool->Sockets, &Socket->Entry, Socket->LocalAddress.Ipv4.sin_port, &Context);
     }
+
     CxPlatRwLockReleaseExclusive(&Pool->Lock);
+
+Error:
```
Comment (nibanks) [inline src/platform/datapath_raw_socket.c:179]: We need to clean up the aux socket on failure. `CxPlatRemoveSocket` isn't called on failure.

**Review by nibanks — APPROVED** [review]

**Comment by nibanks** [comment]
I validated the latest on our Azure VMs.
```
PS C:\test\xdp\bin> .\secnetperf.exe -test:rps -target:172.24.0.4 -conns:1 -requests:1 -request:512 -response:8192
All Connected! Waiting for idle.
Start sending request...
Started!

Result: 11659 RPS, Min: 59, Max: 4068, 50th: 71.000000, 90th: 75.000000, 99th: 90.000000, 99.9th: 481.000000, 99.99th: 2287.000000, 99.999th: 3565.000000, 99.9999th: 4068.000000, StdErr: 0.123326
App Main returning status 0
[0]txInvalidDescriptors: 0
]0]rxDropped: 0
[0]rxInvalidDescriptors: 0
```

File: src/platform/datapath_raw_socket.c (hunk 2 above)
Comment (maolson-msft) [reply src/platform/datapath_raw_socket.c:157]: Why?

Comment (nibanks) [reply src/platform/datapath_raw_socket.c:157]: To also get the right local IP address

Comment (maolson-msft) [reply src/platform/datapath_raw_socket.c:157]: We need to discuss this.

Comment (maolson-msft) [reply src/platform/datapath_raw_socket.c:157]: Discussed- connect sounds good!

**Pull request merged**